### PR TITLE
Beforehand generation of transaction ID when inserting into AO tables.

### PR DESCRIPTION
**Pull request by InnerLife0**
Empty transaction ID can cause errors on dependent entities calculation, such as spgist indexes.
Adding a call to 'GetCurrentTransactionId' like it's done in vanilla Postgres version of 'heap_insert' function.
Each of 4 regression tests is independent.
PR comment and commit message will be updated after discussion.
______________________
QA
- Why not just generate new transaction ID [here](https://github.com/arenadata/gpdb/blob/284466f06aba8c254231ab229961d7d35f430a18/src/backend/executor/nodeModifyTable.c#L479)?
- There are other operators affected. `UPDATE`, `COPY` and, possibly, `ALTER TABLE`.
__
- ~~Why no `UPDATE` regression tests?~~
- ~~AO table update logic reuses insert logic([one](https://github.com/arenadata/gpdb/blob/284466f06aba8c254231ab229961d7d35f430a18/src/backend/access/appendonly/appendonlyam.c#L2576), [two](https://github.com/arenadata/gpdb/blob/284466f06aba8c254231ab229961d7d35f430a18/src/backend/access/aocs/aocsam.c#L1625)).~~
__
- Why no `ALTER TABLE` regression tests?
- I was not able to find working solution to trigger an error inside `ATRewriteTable` [function](https://github.com/arenadata/gpdb/blob/284466f06aba8c254231ab229961d7d35f430a18/src/backend/commands/tablecmds.c#L6274). As I can see, transaction ID correctly generated on previous steps for all cases I found.
__
- Why not place a fix [here](https://github.com/arenadata/gpdb/blob/284466f06aba8c254231ab229961d7d35f430a18/src/backend/executor/nodeModifyTable.c#L564), [here](https://github.com/arenadata/gpdb/blob/284466f06aba8c254231ab229961d7d35f430a18/src/backend/executor/execUtils.c#L1129) or even [here](https://github.com/arenadata/gpdb/blob/284466f06aba8c254231ab229961d7d35f430a18/src/backend/executor/execUtils.c#L1198)?
- It's also possible to place a fix in AO init functions([here](https://github.com/arenadata/gpdb/blob/284466f06aba8c254231ab229961d7d35f430a18/src/backend/access/appendonly/appendonlyam.c#L2595) and [here](https://github.com/arenadata/gpdb/blob/284466f06aba8c254231ab229961d7d35f430a18/src/backend/access/aocs/aocsam.c#L824)). But it seems to be implicitly logical  to place generation of new transaction ID into `insert` function, and more, this solution [inherits](https://github.com/postgres/postgres/blob/REL9_4_STABLE/src/backend/access/heap/heapam.c#L2062) the way of vanilla Postgres version of function `heap_insert`. The drawback of this solution, it triggers the call to `GetCurrentTransactionId` even for non-affected entities, such as [matview](https://github.com/arenadata/gpdb/blob/284466f06aba8c254231ab229961d7d35f430a18/src/backend/commands/matview.c#L575) and [ctas](https://github.com/arenadata/gpdb/blob/284466f06aba8c254231ab229961d7d35f430a18/src/backend/commands/createas.c#L714). At this time, I see it may be preferable to place a fix somewhere inside `ExecInsertIndexTuples`, despite the fact that it contradicts vanilla Postgres way. I want to hear teams opinion.

### CODE REVIEW

File: src/backend/access/aocs/aocsam.c [hunk 1]
```diff
@@ -910,6 +911,12 @@ aocs_insert_values(AOCSInsertDesc idesc, Datum *d, bool *null, AOTupleId *aoTupl
 								   RelationGetRelationName(idesc->aoi_rel));	/* tableName */
 #endif
 
+	/*
+	 * generate new transaction id if necessary, so dependent entities, such as
+	 * spgist indexes, can use it outside aocs_insert_values
+	 */
+	GetCurrentTransactionId();
```
Comment (seryozhasmirnov) [inline src/backend/access/aocs/aocsam.c]: Sometimes` (void) GetCurrentTransactionId();` is used to show that return value is explicitly ignored.
For example, [here](https://github.com/arenadata/gpdb/blob/adb-6.x/src/backend/utils/cache/relcache.c#L3267) or [here](https://github.com/arenadata/gpdb/blob/adb-6.x/src/backend/commands/async.c#L816).
What do you think?

File: src/backend/access/appendonly/appendonlyam.c [hunk 2]
```diff
@@ -2849,6 +2849,12 @@ appendonly_insert(AppendOnlyInsertDesc aoInsertDesc,
 
 	Assert(RelationIsAoRows(relation));
 
+	/*
+	 * generate new transaction id if necessary, so dependent entities, such as
+	 * spgist indexes, can use it outside appendonly_insert
+	 */
+	GetCurrentTransactionId();
```
Comment (seryozhasmirnov) [inline src/backend/access/appendonly/appendonlyam.c]: The same comment as [above](https://github.com/arenadata/gpdb/pull/241/files#r689613609).

File: src/test/regress/expected/create_index.out [hunk 3]
```diff
@@ -3129,3 +3129,70 @@ explain (costs off)
  Optimizer: Postgres query optimizer
 (4 rows)
 
+--
+-- Check spgist index work with AO tables
+--
+-- insert to AO table with spgist index
+create table spgist_ao_test(idx_col int4range)
```
Comment (seryozhasmirnov) [inline src/test/regress/expected/create_index.out:3136]: In this file all other SQL words are in capital (**CREATE** but not **create**).
Should not it be consistent with the rest of the file?

File: src/test/regress/sql/create_index.sql [hunk 4]
```diff
@@ -975,3 +975,89 @@ RESET enable_indexonlyscan;
 
 explain (costs off)
   select * from tenk1 where (thousand, tenthous) in ((1,1001), (null,null));
+
+--
+-- Check spgist index work with AO tables
+--
+-- insert to AO table with spgist index
+create table spgist_ao_test(idx_col int4range)
```
Comment (seryozhasmirnov) [inline src/test/regress/sql/create_index.sql:983]: In this file all other SQL words are in capital (**CREATE** but not **create**).
Should not it be consistent with the rest of the file?

File: src/test/regress/expected/create_index_optimizer.out [hunk 5]
```diff
@@ -3160,3 +3160,70 @@ explain (costs off)
  Optimizer: Pivotal Optimizer (GPORCA) version 3.1.0
 (4 rows)
 
+--
+-- Check spgist index work with AO tables
+--
+-- insert to AO table with spgist index
+create table spgist_ao_test(idx_col int4range)
```
Comment (seryozhasmirnov) [inline src/test/regress/expected/create_index_optimizer.out:3167]: In this file all other SQL words are in capital (**CREATE** but not **create**).
Should not it be consistent with the rest of the file?

File: src/backend/access/aocs/aocsam.c (hunk 1 above)
Comment (InnerLife0) [reply src/backend/access/aocs/aocsam.c]: Will be fixed once we'll make a decision about fix place. Thanks.

File: src/test/regress/sql/create_index.sql (hunk 4 above)
Comment (InnerLife0) [reply src/test/regress/sql/create_index.sql:983]: I don't think it make real sense here. There are no strict rules explaining this format, and more, this set of tests just continues existing vanilla Postgres [test](https://github.com/arenadata/gpdb/pull/241/files/e9dc8c11668fd43e94290573ed6d102ce21d6b3f#diff-6a13a6d37140c67217997c422b6a960bbb089f280bd8c43b0bc6c339a88b606dR976) above, which is all in lowercase.

File: src/test/regress/sql/create_index.sql [hunk 6]
```diff
@@ -975,3 +975,89 @@ RESET enable_indexonlyscan;
...
+insert into spgist_aocs_copy_test
+select int4range(i, i, '[]')
+from generate_series(0,10000) i;
+
+create index spgist_aocs_copy_test_idx on spgist_aocs_copy_test
+using spgist(idx_col);
+
+copy spgist_aocs_copy_test to '/tmp/spgist_test';
+--copy should not fall with spgutils.c:724 assertion or any segment failures
+copy spgist_aocs_copy_test from '/tmp/spgist_test';
```
Comment (darthunix) [inline src/test/regress/sql/create_index.sql]: Please, add a new line here.

File: src/backend/access/aocs/aocsam.c [hunk 7]
```diff
@@ -910,6 +911,12 @@ aocs_insert_values(AOCSInsertDesc idesc, Datum *d, bool *null, AOTupleId *aoTupl
 								   RelationGetRelationName(idesc->aoi_rel));	/* tableName */
 #endif
 
+	/*
+	 * generate new transaction id if necessary, so dependent entities, such as
```
Comment (darthunix) [inline src/backend/access/aocs/aocsam.c]: All the code around starts the sentence from the capital letter and finishes with a dot. I suggest to follow this style.

**Review by darthunix — DISMISSED** [review]
I think that the beginning of `AM_insert()` function is a good place to insert `GetCurrentTransactionId()`. If we take a look at upstream postgres code and inspect its `heap_insert()` implementation, we can see that postgres doesn't pass transaction as a parameter, but rather [generate](https://github.com/postgres/postgres/blob/bad6cef32ca2dd60c8be18f95a7157ced189ec99/src/backend/access/heap/heapam.c#L2063) it with `GetCurrentTransactionId()` in the first line of the function. So here @InnerLife0 uses the same approach.

**Review by darthunix — APPROVED** [review]

**Review by seryozhasmirnov — APPROVED** [review]

**Pull request merged**